File: ios/Packages/OpenUILang/Sources/OpenUILang/AST.swift
import Foundation

/// AST nodes mirroring lang-core `parser/ast.js`
/// (spec/openui-lang.md §5 expression grammar).
/// No `Equatable` conformance on purpose: the runtime never compares AST
/// nodes, and a synthesized/handwritten `==` would only invite accidental
/// (and previously misleading) use.
indirect enum ASTNode {
    case str(String)
    case num(Double)
    case bool(Bool)
    case null
    case arr([ASTNode])
    case obj([(key: String, value: ASTNode)])
    case comp(name: String, args: [ASTNode], mappedProps: [(key: String, value: ASTNode)]?)
    case ref(String)
    case stateRef(String)
    case runtimeRef(name: String, refType: String)
    case binOp(op: String, left: ASTNode, right: ASTNode)
    case unaryOp(op: String, operand: ASTNode)
    case ternary(cond: ASTNode, then: ASTNode, elseNode: ASTNode)
    case member(obj: ASTNode, field: String)
    case index(obj: ASTNode, index: ASTNode)
    case assign(target: String, value: ASTNode)
    case ph(String)

    /// Runtime expression nodes that survive parser lowering.
    var isRuntimeExpr: Bool {
        switch self {
        case .stateRef, .runtimeRef, .binOp, .unaryOp, .ternary, .member, .index, .assign:
            return true
        default:
            return false
        }
    }
}

/// Walk an AST tree, visiting every node (port of `walkAST`).
func walkAST(_ node: ASTNode, _ visit: (ASTNode) -> Void) {
    visit(node)
    switch node {
    case .comp(_, let args, let mappedProps):
        for a in args { walkAST(a, visit) }
        if let mapped = mappedProps {
            for (_, v) in mapped { walkAST(v, visit) }
        }
    case .arr(let els):
        for e in els { walkAST(e, visit) }
    case .obj(let entries):
        for (_, v) in entries { walkAST(v, visit) }
    case .binOp(_, let l, let r):
        walkAST(l, visit)
        walkAST(r, visit)
    case .unaryOp(_, let operand):
        walkAST(operand, visit)
    case .ternary(let c, let t, let e):
        walkAST(c, visit)
        walkAST(t, visit)
        walkAST(e, visit)
    case .member(let obj, _):
        walkAST(obj, visit)
    case .index(let obj, let idx):
        walkAST(obj, visit)
        walkAST(idx, visit)
    case .assign(_, let value):
        walkAST(value, visit)
    default:
        break
    }
}

/// Collect all StateRef ($variable) names referenced within a node,
/// deduplicated preserving first-seen order (port of `collectQueryDeps`).
func collectStateRefs(_ node: ASTNode) -> [String] {
    var seen = Set<String>()
    var out: [String] = []
    walkAST(node) { current in
        if case .stateRef(let n) = current, !seen.contains(n) {
            seen.insert(n)
            out.append(n)
        }
    }
    return out
}

File: android/openui-lang/src/main/kotlin/dev/appless/openuilang/Ast.kt
package dev.appless.openuilang

/**
 * AST nodes mirroring lang-core `parser/ast.js`
 * (spec/openui-lang.md §5 expression grammar).
 *
 * Object/mappedProps entries are kept as ordered `List<Pair<String, AstNode>>`
 * (not a `Map`) because JS builds them with plain-object assignment: insertion
 * order is observable in `$ast` serialization, and duplicate keys are
 * last-wins **at the point of the first insertion** only for `Obj` literals —
 * which the reference implements by `o[k] = v` on a plain object. See
 * [Materialize] / [Evaluator] where the list is folded into an [RtObject].
 */
internal sealed interface AstNode {
    data class Str(val v: String) : AstNode
    data class Num(val v: Double) : AstNode
    data class Bool(val v: Boolean) : AstNode
    data object Null : AstNode
    data class Arr(val els: List<AstNode>) : AstNode
    data class Obj(val entries: List<Pair<String, AstNode>>) : AstNode
    data class Comp(
        val name: String,
        val args: List<AstNode>,
        val mappedProps: List<Pair<String, AstNode>>? = null,
    ) : AstNode

    data class Ref(val n: String) : AstNode
    data class StateRef(val n: String) : AstNode
    data class RuntimeRef(val n: String, val refType: String) : AstNode
    data class BinOp(val op: String, val left: AstNode, val right: AstNode) : AstNode
    data class UnaryOp(val op: String, val operand: AstNode) : AstNode
    data class Ternary(val cond: AstNode, val then: AstNode, val orElse: AstNode) : AstNode
    data class Member(val obj: AstNode, val field: String) : AstNode
    data class Index(val obj: AstNode, val index: AstNode) : AstNode
    data class Assign(val target: String, val value: AstNode) : AstNode
    data class Ph(val n: String) : AstNode
}

/**
 * The node's `k` discriminant — the own property lang-core's duck-typing reads
 * (`isASTNode`, `evaluate`'s switch, the serializer's `isAstNode`).
 */
internal val AstNode.kindTag: String
    get() = when (this) {
        is AstNode.Str -> "Str"
        is AstNode.Num -> "Num"
        is AstNode.Bool -> "Bool"
        is AstNode.Null -> "Null"
        is AstNode.Arr -> "Arr"
        is AstNode.Obj -> "Obj"
        is AstNode.Comp -> "Comp"
        is AstNode.Ref -> "Ref"
        is AstNode.StateRef -> "StateRef"
        is AstNode.RuntimeRef -> "RuntimeRef"
        is AstNode.BinOp -> "BinOp"
        is AstNode.UnaryOp -> "UnaryOp"
        is AstNode.Ternary -> "Ternary"
        is AstNode.Member -> "Member"
        is AstNode.Index -> "Index"
        is AstNode.Assign -> "Assign"
        is AstNode.Ph -> "Ph"
    }

/** Runtime expression nodes that survive parser lowering (`isRuntimeExpr`). */
internal val AstNode.isRuntimeExpr: Boolean
    get() = when (this) {
        is AstNode.StateRef, is AstNode.RuntimeRef, is AstNode.BinOp, is AstNode.UnaryOp,
        is AstNode.Ternary, is AstNode.Member, is AstNode.Index, is AstNode.Assign,
        -> true

        else -> false
    }

/** Walk an AST tree, visiting every node (port of `walkAST`). */
internal fun walkAst(node: AstNode, visit: (AstNode) -> Unit) {
    visit(node)
    when (node) {
        is AstNode.Comp -> {
            for (a in node.args) walkAst(a, visit)
            node.mappedProps?.forEach { walkAst(it.second, visit) }
        }

        is AstNode.Arr -> node.els.forEach { walkAst(it, visit) }
        is AstNode.Obj -> node.entries.forEach { walkAst(it.second, visit) }
        is AstNode.BinOp -> {
            walkAst(node.left, visit)
            walkAst(node.right, visit)
        }

        is AstNode.UnaryOp -> walkAst(node.operand, visit)
        is AstNode.Ternary -> {
            walkAst(node.cond, visit)
            walkAst(node.then, visit)
            walkAst(node.orElse, visit)
        }

        is AstNode.Member -> walkAst(node.obj, visit)
        is AstNode.Index -> {
            walkAst(node.obj, visit)
            walkAst(node.index, visit)
        }

        is AstNode.Assign -> walkAst(node.value, visit)
        else -> Unit
    }
}

/**
 * Collect all `StateRef` (`$variable`) names referenced within a node,
 * deduplicated preserving first-seen order (port of `collectQueryDeps`).
 *
 * A `LinkedHashSet<String>` is exactly a JS `Set` of strings here: JVM
 * `String.equals`/`hashCode` are UTF-16 code-unit based, so no `JSKey`
 * wrapper (as in the Swift port) is needed.
 */
internal fun collectStateRefs(node: AstNode): List<String> {
    val out = LinkedHashSet<String>()
    walkAst(node) { if (it is AstNode.StateRef) out.add(it.n) }
    return out.toList()
}
